lib/src/rules/always_use_required_for_non_null_named_parameter.dart

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.always_use_required_for_non_null_named_parameter;
+
+import 'package:analyzer/analyzer.dart';
+import 'package:analyzer/dart/ast/ast.dart' show AstVisitor, TypedLiteral;
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const desc = 'Use @required.';
+
+const details = '''
+**DO** specify `@required` on named parameter without default value on which an
+assert(param != null) is done.
+
+**GOOD:**
+```
+m1({@required a}) {
+  assert(a != null);
+}
+
+m2({a: 1}) {
+  assert(a != null);
+}
+```
+
+**BAD:**
+```
+m1({a}) {
+  assert(a != null);
+}
+```
+
+NOTE: Only asserts at the start of the bodies will be taken into account.
+''';
+
+/// The name of `meta` library, used to define analysis annotations.
+String _META_LIB_NAME = "meta";
+
+/// The name of the top-level variable used to mark a required named parameter.
+String _REQUIRED_VAR_NAME = "required";
+
+bool _isRequired(Element element) =>
+    element is PropertyAccessorElement &&
+    element.name == _REQUIRED_VAR_NAME &&
+    element.library?.name == _META_LIB_NAME;
+
+class AlwaysUseRequiredForNonNullNamedParameter extends LintRule {
+  AlwaysUseRequiredForNonNullNamedParameter()
+      : super(
+            name: 'always_use_required_for_non_null_named_parameter',
+            description: desc,
+            details: details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  Visitor(this.rule);
+
+  void checkLiteral(TypedLiteral literal) {
+    if (literal.typeArguments == null) {
+      rule.reportLintForToken(literal.beginToken);
+    }
+  }
+
+  @override
+  visitFormalParameterList(FormalParameterList node) {
```

Comment:
The performance might be fine as is, but you could make this more efficient by gathering the names of all of the variables whose null-ness is asserted before iterating over the list of required parameters. For methods with more than one named parameter this would reduce the amount of duplicated work.